### PR TITLE
bpo-45756: Fix mock triggers dynamic lookup via the descriptor protocol.

### PR DESCRIPTION
**Pull request by hongweipeng**
https://bugs.python.org/issue45756

### CODE REVIEW

**Review by sobolevn — COMMENTED** [review]
Isn't it the same as https://github.com/python/cpython/pull/29901 ?

**Comment by hongweipeng** [comment]
> Isn't it the same as #29901 ?

I think this is a good example of using `inspect.getmembers_static`.

File: Lib/unittest/mock.py [hunk 1]
```diff
@@ -496,9 +496,8 @@ def _mock_add_spec(self, spec, spec_set, _spec_as_instance=False,
         _spec_signature = None
         _spec_asyncs = []
 
-        for attr in dir(spec):
-            if iscoroutinefunction(getattr(spec, attr, None)):
-                _spec_asyncs.append(attr)
+        for attr, _ in inspect.getmembers_static(spec, iscoroutinefunction):
```
Comment (JelleZijlstra) [inline Lib/unittest/mock.py:499]: What about async classmethods? Would those still get recognized as coroutines?

**Comment by carljm** [comment]
Thanks for the PR! I think we should consolidate the discussion to #22209. If we do this, we should prefer continuing to use `dir` and using `getattr_static` (as done in #22209), rather than switching to `getmembers_static`, because switching away from `dir` causes extra avoidable backward-incompatibility.

Also, I think the async-classmethod case is not fixable using `getmembers_static`, but is fixable via `getattr_static`.

Closing as duplicate of #22209.

**Pull request closed without merging**